File: 29. Divide Two Integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        '''
        Consider a 32-bit machine.
        The only case to overflow is INT_MIN / -1
        and to prevent abs() overflow, we use negative numbers
        and the method is to find the closest 2^n*divisor and subtract that
        and we add n to result.
        '''
        if dividend == -2**31 and divisor == -1:
            return 2**31 -1
        sign = (dividend < 0) == (divisor < 0)
        if dividend > 0:
            dividend = 0 - dividend
        if divisor > 0:
            divisor = 0 - divisor
        res = 0
        while dividend <= divisor:
            temp = 1
            div = divisor
            while dividend <= div:
                div <<= 1
                temp <<= 1
            res += (temp>>1)
            dividend -= (div>>1)
        if sign:
            return res
        else:
            return 0 - res

File: 29. Divide Two Integers.java
class Solution {
    public int divide(int dividend, int divisor) {
        // Careful for overflow and edge cases
        // Thought: find the most 2^k * divisor and substract that
        if (dividend == (Integer.MIN_VALUE) && divisor == -1) {
            return Integer.MAX_VALUE;
        } else if (dividend == (Integer.MIN_VALUE) && divisor == 1) {
            return Integer.MIN_VALUE;
        }
        boolean samesign = !((dividend ^ divisor) < 0);
        if (dividend > 0) {
            dividend = 0 - dividend;
        }
        if (divisor > 0) {
            divisor = 0 - divisor;
        }
        int res = 0;
        while (dividend <= divisor) {
            int d = divisor;
            int count = 0;
            while (d >= dividend) {
                if (d < (Integer.MIN_VALUE) >> 1) {
                    break;
                }
                d = d << 1;
                count++;
            }
            if (count > 0) {
                dividend = dividend - (d >> 1);
                res += Math.pow(2, count-1);
            } else {
                dividend = dividend - d;
                res += 1;
            }
            
        }
        if (!samesign) {
            res = 0 - res;
        }
        return res;
    }
}